### PR TITLE
Modified circle finder, target locator, intrinsic calibration, covariance computation etc

### PR DESCRIPTION
**Pull request by drchrislewis**
The main improvements include
   A modified circle grid finder that understands orientation via a slightly larger circle at the origin
   A target finder package
   An intrinsic calibration package
   The ability to ask for covariance after the solver completes
   New calling procedure which indicates an acceptable error, and returns the per/observation error

### CODE REVIEW

File: industrial_extrinsic_cal/include/industrial_extrinsic_cal/ceres_costs_utils.hpp [hunk 1]
```diff
@@ -1253,17 +1439,180 @@ namespace industrial_extrinsic_cal
...
+    static ceres::CostFunction* Create(const double o_x, const double o_y, const double c_dia)
+    {
+      return (new ceres::AutoDiffCostFunction<CircleTargetCameraReprjErrorWithDistortion, 2, 6, 9, 3>(new CircleTargetCameraReprjErrorWithDistortion(o_x, o_y, c_dia)));
+    }
+    double ox_; /** observed x location of object in image */
+    double oy_; /** observed y location of object in image */
+    double circle_diameter_; //** diameter of circle being observed */
+  };
+
+  /******************I'm working on this one now *************************************************************/
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/include/industrial_extrinsic_cal/ceres_costs_utils.hpp]: Remove comments.

File: industrial_extrinsic_cal/include/industrial_extrinsic_cal/circle_detector.hpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#ifndef CIRCLEDETECTOR_HPP
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/include/industrial_extrinsic_cal/circle_detector.hpp]: Is than an openCV file?  It's missing it's license.

File: industrial_extrinsic_cal/include/industrial_extrinsic_cal/camera_yaml_parser.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include <iostream>
+#include <industrial_extrinsic_cal/basic_types.h>
+#include <industrial_extrinsic_cal/camera_definition.h>
+#include <industrial_extrinsic_cal/transform_interface.hpp>
+#include <industrial_extrinsic_cal/trigger.h>
+
+
+namespace industrial_extrinsic_cal {
+
+  using std::string;
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/include/industrial_extrinsic_cal/camera_yaml_parser.h]: Typically we don't use `using` in a header.  This pollutes any file that might import the header.  These are "scoped" so maybe it's not an issue.  Thoughts?

File: industrial_extrinsic_cal/include/industrial_extrinsic_cal/ceres_costs_utils.hpp [hunk 4]
```diff
@@ -1906,7 +2255,10 @@ namespace industrial_extrinsic_cal
       T cy = T(cy_);
       T ox = T(ox_);
       T oy = T(oy_);
-      cameraCircResidual(camera_point, circle_diameter, R_optical_to_target, fx, fy,cx,cy, ox, oy, resid);
+      // cameraPntResidual(camera_point, fx, fy,cx,cy, ox, oy, residual);
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/include/industrial_extrinsic_cal/ceres_costs_utils.hpp]: Remove commented out code.

File: industrial_extrinsic_cal/include/industrial_extrinsic_cal/camera_yaml_parser.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  /** @brief parses a Pose6d
+   *   @param node, the yaml node 
+   *   @return Pose6d parsed
+   **/
+  Pose6d parsePose(const Node &node);
+  /** @brief parse all the cameras, the main function of this module
+   *   @param camera_input_file the stream from which to parse a vector of cameras
+   *   @param returned vector of shared pointers to cameras
+   **/
+  void parseCameras(ifstream &cameras_input_file,vector<shared_ptr<Camera> >& cameras);
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/include/industrial_extrinsic_cal/camera_yaml_parser.h]: What if the parse methods fails?

File: industrial_extrinsic_cal/include/industrial_extrinsic_cal/points_yaml_parser.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+#include <iostream>
+#include <industrial_extrinsic_cal/basic_types.h>
+
+namespace industrial_extrinsic_cal {
+
+  /** @brief parse a list of points from a file
+   *  @param points_input_file, the stream from which to parse the points
+   *  @param points the returned vector of points
+   */
+  void parsePoints(std::ifstream &points_input_file, std::vector<industrial_extrinsic_cal::Point3d> &points);
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/include/industrial_extrinsic_cal/points_yaml_parser.h]: What if the parse call fails?

File: industrial_extrinsic_cal/include/industrial_extrinsic_cal/targets_yaml_parser.h [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#include <industrial_extrinsic_cal/basic_types.h>
+#include <industrial_extrinsic_cal/target.h>
+
+namespace industrial_extrinsic_cal {
+  // prototypes
+  /** @brief parse targets from a yaml file
+   *    @param target_input_file the stream from which to parse the targets
+   *    @param targets vector of shared pointers to targets parsed
+   */
+  void parse_targets(std::ifstream &targets_input_file, std::vector<boost::shared_ptr<Target> > & targets);
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/include/industrial_extrinsic_cal/targets_yaml_parser.h]: What if this fails?

File: industrial_extrinsic_cal/src/calibration_job_definition.cpp [hunk 8]
```diff
@@ -32,10 +32,81 @@ using std::string;
...
+	  fprintf(fp,"%s ", odl[i].items_[j].camera_name_.c_str());
+	}
+	camera_name =  odl[i].items_[j].camera_name_;
+      }
+      fprintf(fp, "]\n");
+    }
+  }
+  fclose(fp);
+}
+  CovarianceRequestType intToCovRequest(int request)
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/src/calibration_job_definition.cpp]: This method is extremely fragile.  Changes to the header where this is defined will break this method (although nobody will know).

File: industrial_extrinsic_cal/src/calibration_job_definition.cpp [hunk 9]
```diff
@@ -798,12 +921,22 @@ namespace industrial_extrinsic_cal
 
   bool CalibrationJob::runOptimization()
   {
-    int total_observations =0;
+    // REMOVE ONCE DONE WITH NIST
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/src/calibration_job_definition.cpp]: Is there a way around this?  If not, add an issue and make sure it's addressed.

File: industrial_extrinsic_cal/src/nodes/mutable_joint_state_publisher.cpp [hunk 10]
```diff
@@ -40,13 +40,13 @@ namespace industrial_extrinsic_cal
...
-    get_server_ = nh_.advertiseService("/get_mutable_joint_states", &MutableJointStatePublisher::getCallBack,this);
-    store_server_ = nh_.advertiseService("/store_mutable_joint_states", &MutableJointStatePublisher::storeCallBack, this);
+    set_server_ = nh_.advertiseService( "set_mutable_joint_states", &MutableJointStatePublisher::setCallBack, this);
+    get_server_ = nh_.advertiseService("get_mutable_joint_states", &MutableJointStatePublisher::getCallBack,this);
+    store_server_ = nh_.advertiseService("store_mutable_joint_states", &MutableJointStatePublisher::storeCallBack, this);
 
     // advertise the topic for continious publication of all the mutable joint states
     int queue_size = 10;
-    joint_state_pub_ = nh_.advertise<sensor_msgs::JointState>("/mutable_joint_states", queue_size);
+    joint_state_pub_ = nh_.advertise<sensor_msgs::JointState>("mutable_joint_states", queue_size);
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/src/nodes/mutable_joint_state_publisher.cpp]: +1 to this change ;)

File: industrial_extrinsic_cal/src/nodes/nist_anal.cpp [hunk 11]
```diff
@@ -0,0 +1,1121 @@
+#include <stdlib.h>
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/src/nodes/nist_anal.cpp]: License?  Could this file be named anything else?

File: industrial_extrinsic_cal/src/nodes/nist_anal.cpp [hunk 12]
```diff
@@ -0,0 +1,1121 @@
...
+#include <ostream>
```
Comment (shaun-edwards) [inline industrial_extrinsic_cal/src/nodes/nist_anal.cpp]: Stopping here...for the night.

**Comment by shaun-edwards** [comment]
@drchrislewis, any chance the NIST specific files/methods/nodes can be moved into a separate package?

File: industrial_extrinsic_cal/src/calibration_job_definition.cpp [hunk 13]
```diff
@@ -32,10 +32,81 @@ using std::string;
...
 using boost::make_shared;
 using ceres::CostFunction;
+using industrial_extrinsic_cal::covariance_requests::CovarianceRequestType;
 
 namespace industrial_extrinsic_cal
 {
-
+  
+  /** @brief, a function used for debugging, and generating output for post processing */
+#define WRITE_DEBUG true
```
Comment (drchrislewis) [inline industrial_extrinsic_cal/src/calibration_job_definition.cpp]: These hacks are scheduled to be fixed in next pull request, ignore for now

File: industrial_extrinsic_cal/src/calibration_job_definition.cpp (hunk 9 above)
Comment (drchrislewis) [inline industrial_extrinsic_cal/src/calibration_job_definition.cpp]: will fix in next pull

File: industrial_extrinsic_cal/src/nodes/nist_anal.cpp (hunk 11 above)
Comment (drchrislewis) [inline industrial_extrinsic_cal/src/nodes/nist_anal.cpp]: renamed to not be so anal

**Comment by drchrislewis** [comment]
Accepting this major pull, but some older applications may break. 
Expect more pulls soon

**Pull request merged**